### PR TITLE
Bug fix: expire invalid access tokens

### PR DESCRIPTION
**Pull request by negzi**
A fix for actually expiring the access tokens with  a time limit.

### CODE REVIEW

**Comment by erikjohnston** [comment]
@matrixbot ok to test

**Comment by negzi** [comment]
@erikjohnston 
I looked at the Jenkins job console output and do not think the failure in Sytes Dendron is related to what I have changed. Could you please take a look at this?

File: synapse/api/auth.py [hunk 1]
```diff
@@ -629,7 +629,13 @@ def get_user_by_access_token(self, token, rights="access"):
         except AuthError:
             # TODO(daniel): Remove this fallback when all existing access tokens
             # have been re-issued as macaroons.
+            if self.hs.config.expire_access_token:
+                raise AuthError(
+                    401,
+                    "access token has expired"
+                )
```
Comment (erikjohnston) [inline synapse/api/auth.py]: We don't actually know if the `AuthError` was due to the macaroon having expired at this point. 

What was the reason for raising here? To give an appropriate error message? Or stopping it from falling back to checking if the access token is in the database?

Comment (negzi) [inline synapse/api/auth.py]: I agree that the message in Autherror should be changed. But the problem with this function is it never raises if the access token is expired and not valid anymore, it just looks in database and returns the user anyway with the fallback. That is why I would like to raise if the access token is expired

File: tests/api/test_auth.py [hunk 2]
```diff
@@ -293,3 +293,41 @@ def test_get_user_from_macaroon_expired(self):
...
         self.assertIn("Invalid macaroon", cm.exception.msg)
+
+    @defer.inlineCallbacks
+    def test_get_user_from_expired_access_token(self):
+        self.get_user_from_macaroon = Mock(return_value=AuthError)
+        self.hs.config.expire_access_token = True
+        with self.assertRaises(AuthError) as cm:
+            yield self.auth.get_user_by_access_token('alsdghsdfvedshoeng')
+        self.assertEqual(401, cm.exception.code)
+        self.assertIn("access token has expired", cm.exception.msg)
```
Comment (erikjohnston) [inline tests/api/test_auth.py]: I wouldn't bother testing if the msg string is correct.

File: synapse/api/auth.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/api/auth.py]: I've taken a look at the macaroon stuff and it doesn't appear to be easy to tell which caveats have failed. I'd be sorely tempted to just reraise the exception if `config.expire_access_token` is enabled (which would then effectively disable support for creating non-macaroon access tokens). i.e.:

```python
except AuthError:
    if self.hs.config.expire_access_token:
        raise
    ret = yield self._look_up_user_by_access_token(token)

**Comment by erikjohnston** [comment]
> I looked at the Jenkins job console output and do not think the failure in Sytes Dendron is related to what I have changed. Could you please take a look at this?

Looks like that test has been misconfigured, ignore it and I'll take a look at it in a bit

File: synapse/api/auth.py (hunk 1 above)
Comment (negzi) [inline synapse/api/auth.py]: Yes, I agree you cannot really tell which Caveat(restriction) is failing here. But I didn't understand your last comment very well. Could you please explain a bit. Do you mean you would just raise here with no error message?

File: tests/api/test_auth.py (hunk 2 above)
Comment (negzi) [inline tests/api/test_auth.py]: Sure I will remove this actually forgot to do it, wanted to know if the exception raises for invalid token and for my own assurance :)

File: synapse/api/auth.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/api/auth.py]: `raise` by itself will raise the previous exception raised, so:

```python
try:
    ...
except Exception:
    raise
```

is equivalent to:

```python
try:
    ...
except Exception as e:
    raise e
```

(which would be equivalent to not catching the exception).

Hence:

```python
except AuthError:
    if self.hs.config.expire_access_token:
        raise
    ret = yield self._look_up_user_by_access_token(token)
```

will effectively only catch the exception if `self.hs.config.expire_access_token` is `True`

(You need to be a little bit careful with using `raise` versus `raise e` if you call a function before re-raising the exception, as the function itself may raise and catch an exception.)

Comment (negzi) [inline synapse/api/auth.py]: Thanks for the awesome explanation :) I will fix this now

**Comment by negzi** [comment]
@erikjohnston 
Please let me know if I need to add more test cases. Thanks!

**Comment by erikjohnston** [comment]
This looks great, thanks! Just need you to sign it off and then I'll merge it

**Pull request merged**